// @colyseus/schema/lib/codegen/languages/lua.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generate = void 0;
var types_1 = require("../types");
/**
    TODO:
    - Support inheritance
    - Support importing Schema dependencies
*/
var typeMaps = {
    "string": "string",
    "number": "number",
    "boolean": "boolean",
    "int8": "number",
    "uint8": "number",
    "int16": "number",
    "uint16": "number",
    "int32": "number",
    "uint32": "number",
    "int64": "number",
    "uint64": "number",
    "float32": "number",
    "float64": "number",
};
var distinct = function (value, index, self) { return self.indexOf(value) === index; };
function generate(context, options) {
    return context.classes.map(function (klass) { return ({
        name: klass.name + ".lua",
        content: generateClass(klass, options.namespace, context.classes)
    }); });
}
exports.generate = generate;
function generateClass(klass, namespace, allClasses) {
    var allRefs = [];
    klass.properties.forEach(function (property) {
        var type = property.type;
        // keep all refs list
        if ((type === "ref" || type === "array" || type === "map")) {
            allRefs.push(property);
        }
    });
    // TOOD: inheritance
    return "".concat((0, types_1.getCommentHeader)().replace(/\/\//mg, "--"), "\n\nlocal schema = require 'colyseus.serialization.schema.schema'\n").concat(allRefs.
        filter(function (ref) { return ref.childType && typeMaps[ref.childType] === undefined; }).
        map(function (ref) { return ref.childType; }).
        concat((0, types_1.getInheritanceTree)(klass, allClasses, false).map(function (klass) { return klass.name; })).
        filter(distinct).
        map(function (childType) { return "local ".concat(childType, " = require '").concat((namespace ? "".concat(namespace, ".") : '')).concat(childType, "'"); }).
        join("\n"), "\n\nlocal ").concat(klass.name, " = schema.define({\n").concat(klass.properties.map(function (prop) { return generatePropertyDeclaration(prop); }).join(",\n"), ",\n    [\"_fields_by_index\"] = { ").concat(klass.properties.map(function (prop) { return "\"".concat(prop.name, "\""); }).join(", "), " },\n})\n\nreturn ").concat(klass.name, "\n");
    // ["on_change"] = function(changes)
    //     -- on change logic here
    // end,
    // ["on_add"] = function()
    //     -- on add logic here
    //  end,
    // ["on_remove"] = function()
    //     -- on remove logic here
    // end,
}
function generatePropertyDeclaration(prop) {
    var typeArgs;
    if (prop.childType) {
        var isUpcaseFirst = prop.childType.match(/^[A-Z]/);
        if (isUpcaseFirst) {
            typeArgs += "".concat(prop.childType);
        }
        else {
            typeArgs += "\"".concat(prop.childType, "\"");
        }
        if (prop.type === "ref") {
            typeArgs = (isUpcaseFirst)
                ? "".concat(prop.childType)
                : "\"".concat(prop.childType, "\"");
        }
        else {
            typeArgs = (isUpcaseFirst)
                ? "{ ".concat(prop.type, " = ").concat(prop.childType, " }")
                : "{ ".concat(prop.type, " = \"").concat(prop.childType, "\" }");
        }
    }
    else {
        typeArgs = "\"".concat(prop.type, "\"");
    }
    return "    [\"".concat(prop.name, "\"] = ").concat(typeArgs);
}
// function generatePropertyInitializer(prop: Property) {
//     let initializer = "";
//     if(prop.type === "ref") {
//         initializer = `new ${prop.childType}()`;
//     } else if(prop.type === "array") {
//         initializer = `new schema.ArraySchema()`;
//     } else if(prop.type === "map") {
//         initializer = `new schema.MapSchema()`;
//     }
//     return `this.${prop.name} = ${initializer}`;
// }
